vllm/config.py

```diff
@@ -275,6 +300,160 @@ def get_num_layers(self, parallel_config: "ParallelConfig") -> int:
         total_num_hidden_layers = self.hf_text_config.num_hidden_layers
         return total_num_hidden_layers // parallel_config.pipeline_parallel_size
 
+    def _get_and_verify_dtype(
+        self,
+        config: PretrainedConfig,
+        dtype: Union[str, torch.dtype],
+    ) -> torch.dtype:
+        # NOTE: getattr(config, "torch_dtype", torch.float32) is not correct
+        # because config.torch_dtype can be None.
+        config_dtype = getattr(config, "torch_dtype", None)
+        if config_dtype is None:
+            config_dtype = torch.float32
+
+        if isinstance(dtype, str):
+            dtype = dtype.lower()
+            if dtype == "auto":
+                if config_dtype == torch.float32:
+                    # Following the common practice, we use float16 for float32
+                    # models.
+                    torch_dtype = torch.float16
+                else:
+                    torch_dtype = config_dtype
+            else:
+                if dtype not in _STR_DTYPE_TO_TORCH_DTYPE:
+                    raise ValueError(f"Unknown dtype: {dtype}")
+                torch_dtype = _STR_DTYPE_TO_TORCH_DTYPE[dtype]
+        elif isinstance(dtype, torch.dtype):
+            torch_dtype = dtype
+        else:
+            raise ValueError(f"Unknown dtype: {dtype}")
+
+        if is_hip() and torch_dtype == torch.float32:
+            rocm_supported_dtypes = [
+                k for k, v in _STR_DTYPE_TO_TORCH_DTYPE.items()
+                if (k not in _ROCM_NOT_SUPPORTED_DTYPE)
+            ]
+            raise ValueError(f"dtype '{dtype}' is not supported in ROCm. "
+                             f"Supported dtypes are {rocm_supported_dtypes}")
+
+        # Verify the dtype.
+        if torch_dtype != config_dtype:
+            if torch_dtype == torch.float32:
+                # Upcasting to float32 is allowed.
+                pass
+            elif config_dtype == torch.float32:
+                # Downcasting from float32 to float16 or bfloat16 is allowed.
+                pass
+            else:
+                # Casting between float16 and bfloat16 is allowed with warning.
+                logger.warning("Casting %s to %s.", config_dtype, torch_dtype)
+
+        return torch_dtype
```

Comment:
@robertgshaw2-neuralmagic Do you still remember why you moved these methods from a global function to a method of `ModelConfig`?

Reply:
ref #4846

Reply:
I noticed the function was only used by the `ModelConfig` and no where else, so I thought it made sense to encapulate